File: generated_code/cpp/solution_1_1.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        // Map to store the value of an element and its corresponding index
        std::unordered_map<int, int> num_to_index;
        
        // Iterate through the array once: O(n) Time Complexity
        for (int i = 0; i < nums.size(); ++i) {
            int complement = target - nums[i];
            
            // Check if the complement exists in the map
            if (num_to_index.find(complement) != num_to_index.end()) {
                // If found, return the indices of the complement and the current element
                return {num_to_index[complement], i};
            }
            
            // Otherwise, store the current element and its index in the map
            // Space Complexity: O(n) to store elements in the worst case
            num_to_index[nums[i]] = i;
        }
        
        // The problem guarantees exactly one solution, so this line is theoretically unreachable.
        return {};
    }
};

File: generated_code/py/solution_1_1.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Map to store the value and its corresponding index
        # Time Complexity: O(N) since we traverse the list containing N elements only once.
        # Space Complexity: O(N) to store the elements in the hash map.
        num_to_index = {}
        
        for index, num in enumerate(nums):
            complement = target - num
            
            # If the complement exists in the map, we found our pair
            if complement in num_to_index:
                return [num_to_index[complement], index]
            
            # Otherwise, store the current number and its index in the map
            num_to_index[num] = index
            
        # The problem guarantees exactly one solution, so we don't need a fallback return